### DISCUSSION TITLE
Max Packet Size

### DISCUSSION THREAD

**Discussion by tebrown** [jgromes/RadioLib#1291] [Q&A]
Hi All,

  Thanks to the help of folks here, I was able to get join requests working.  Right now, my max packet size though is 9 bytes.  I am looking to send closer to 32 bytes of data across in a single payload.  I would have expected that the smallest maximum size based on this chart would be ~50 bytes.  
![475417385baf2c87fe595c8c960b3667d76a6c60](https://github.com/user-attachments/assets/045ccc43-2008-46ca-a89e-e1658807430f)

```
Reconnecting to /dev/cu.usbmodem111401 ....................      Connected!
10:04:43.077 > Recalling LoRaWAN nonces & session
10:04:43.080 > Join ('login') to the LoRaWAN Network
10:04:48.667 > Saving nonces to flash
10:04:48.672 > Joined
10:04:49.673 > Sending uplink
10:04:49.673 > Max packet size: 9
10:04:49.677 > Sending uplink
10:04:51.198 > Received a downlink
10:04:51.198 > <MAC commands only>
10:04:51.198 > [LoRaWAN] RSSI:          -80.00 dBm
10:04:51.198 > [LoRaWAN] SNR:           6.50 dB
10:04:51.198 > [LoRaWAN] Event information:
10:04:51.199 > [LoRaWAN] Confirmed:     0
10:04:51.199 > [LoRaWAN] Confirming:    0
10:04:51.199 > [LoRaWAN] Datarate:      10
10:04:51.199 > [LoRaWAN] Frequency:     926.900 MHz
10:04:51.199 > [LoRaWAN] Frame count:   0
10:04:51.199 > [LoRaWAN] Port:          0
10:04:51.199 > [LoRaWAN] Time-on-air:   371 ms
10:04:51.199 > [LoRaWAN] Rx window:     1
10:04:51.200 > FcntUp: 0
10:04:51.200 > Sleeping
10:04:51.200 > 
```

**Comment by StevenCellist** [comment]
@tebrown so what's the question here? :)

**Reply by StevenCellist** [reply]
Never mind - I know see the problem. The point is: dwell time limitations! (Yay...). In the US, you are limited to 400ms maximum time-on-air per packet. [Here's a handy tool](https://avbentem.github.io/airtime-calculator/ttn/us915) to see what that means. Best is to keep ADR enabled such that you can hopefully move to a faster datarate and get a larger allowed packet size.

**Reply by tebrown** [reply]
You are so responsive!  Where can I make a donation?  Seriously!

If I set the uplink to 16 octets, it fails with a -4 (RADIOLIB_ERR_PACKET_TOO_LONG), which I would expect since it is returning the max packet size is 9.  Is the right answer to fragment this, but then I need to use one of the fields for a key or sequence number, which is not ideal.  I think I am missing something.

**Reply by StevenCellist** [reply]
Given that shipping costs are quite high across the Atlantic, you're welcome to make a donation to my student project such that I can order a couple more parts for our sensor boxes :) LMK!

The best answer is to move to a faster datarate: you can get up to 53 octets at DR1 instead of DR0 (although sometimes you may see a few bytes less due to network octets piggybacking onto the frame). You can force this by calling `node.setDatarate(1)`, although we'd usually recommend using ADR (Adaptive Data Rate) which allows the network to optimize the datarate for your device (ADR is enabled by default). If you have decent coverage, you really don't need DR0 or DR1 and can save some battery life by using DR2 even DR3. The network can control this automatically by looking at the RSSI and SNR of the uplinks. (Refer to the TTN Learn pages for more info!)
Usually it takes a couple of uplinks for ADR to kick in if needed, so maybe first send a few 'empty'-ish uplinks. Or disable ADR and set a manual datarate, that's fine, but then you are responsible for being a good neighbor by using the fastest datarate that still gets your messages through.

**Reply by tebrown** [reply]
Awesome!  Thanks for the assistance.  If you can send me a link to your student project, I'll  make a donation. Thank you!

**Reply by StevenCellist** [reply]
You can find information [on our website](https://www.meetjeleefomgeving.nl/contributing/). Thank you very much in advance!

**Reply by StevenCellist** [reply]
@tebrown you're a LEGEND - THANK YOU! ❤️

**Reply by tebrown** [reply]
I disagree.. You are!  Thank you very, very much!

**Reply by StevenCellist** [reply]
![PXL_20241105_132600374](https://github.com/user-attachments/assets/a81876d7-2eeb-4ef4-8d24-7f5fb82ffc55)
Donation has been cashed: got displays for a whole bunch of boxes now, woohoo 🎉 
Thank you!!

**Reply by tebrown** [reply]
Awesome!  I am so glad I could help!
